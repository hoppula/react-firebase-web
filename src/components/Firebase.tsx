import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"

export interface FirebaseProps {
  readonly apiKey: string
  readonly projectId: string
  readonly databaseURL?: string
}

export class Firebase extends React.Component<FirebaseProps, {}> {
  static propTypes = {
    apiKey: PropTypes.string.isRequired,
    databaseURL: PropTypes.string,
    projectId: PropTypes.string.isRequired
  }

  static childContextTypes = {
    firebase: PropTypes.object
  }

  firebase: firebase.app.App

  constructor(props: FirebaseProps) {
    super(props)
    const { apiKey, projectId, databaseURL } = props
    this.firebase = firebase.initializeApp({
      apiKey,
      authDomain: `${projectId}.firebaseapp.com`,
      databaseURL: databaseURL
        ? databaseURL
        : `https://${projectId}.firebaseio.com`,
      storageBucket: `${projectId}.appspot.com`
    })
  }

  getChildContext() {
    return {
      firebase: this.firebase
    }
  }

  render() {
    return React.Children.only(this.props.children)
  }
}

export default Firebase
