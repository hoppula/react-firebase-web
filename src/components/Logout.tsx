import * as PropTypes from "prop-types"
import * as React from "react"
import { FirebaseContext } from "../types"

export interface LogoutProps {
  readonly children: (logout: () => Promise<any>) => React.ReactElement<any>
}

export class Logout extends React.Component<LogoutProps, {}> {
  static propTypes = {
    children: PropTypes.func.isRequired
  }

  static contextTypes = {
    firebase: PropTypes.object.isRequired
  }

  context: FirebaseContext

  logout = () => {
    return this.context.firebase.auth().signOut()
  }

  render() {
    return React.Children.only(this.props.children(this.logout))
  }
}

export default Logout
