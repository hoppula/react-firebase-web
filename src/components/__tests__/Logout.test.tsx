import React from "react"
import renderer from "react-test-renderer"
import { Firebase } from "../Firebase"
import Logout from "../Logout"

import { initializeMockSDK } from "./setup"
const mocksdk = initializeMockSDK()

jest.mock("firebase/app", () => {
  return {
    initializeApp: (...args) => mocksdk.initializeApp(...args)
  }
})

describe("Logout", () => {
  test("currentUser gets emptied when provided render prop function is used", () => {
    let component = renderer.create(
      <Firebase apiKey="firebase-test" projectId="firebase-test">
        <Logout children={logout => <button onClick={logout}>Logout</button>} />
      </Firebase>
    )
    const auth = component.getInstance().firebase.auth()
    let tree = component.toJSON()

    auth.changeAuthState({ uid: "test", displayName: "Test User" })
    auth.flush()
    expect(auth.currentUser.uid).toBe("test")
    expect(tree).toMatchSnapshot()

    const button = component.root.findByType("button")
    button.props.onClick()
    auth.flush()

    expect(auth.currentUser).toBeNull()
  })
})
