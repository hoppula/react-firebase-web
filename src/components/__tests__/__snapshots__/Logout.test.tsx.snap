// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout currentUser gets emptied when provided render prop function is used 1`] = `
<button
  onClick={[Function]}
>
  Logout
</button>
`;
